### ISSUE TITLE
Use `pyproject.toml` to replace `setup.py` and `setup.cfg`

### ISSUE THREAD

**Issue by laymonage** [wagtail/wagtail#12482]
### Is your proposal related to a problem?

The (not so) new recommendation for packages is to use `pyproject.toml` to define the configuration for packaging tools, as well as other tools like linters.

https://packaging.python.org/en/latest/guides/writing-pyproject-toml/

### Describe the solution you'd like

We should migrate our `setup.py` and `setup.cfg` and consolidate it into a new `pyproject.toml` file. This is good for future-proofing, as well as preventing issues with editable installs like in https://github.com/wagtail/gitpod-wagtail-develop/pull/8.

### Describe alternatives you've considered

Keep using `setup.py` and `setup.cfg`.

### Additional context

Our cookiecutter for packages already use `pyproject.toml`: https://github.com/wagtail/cookiecutter-wagtail-package/blob/main/%7B%7B%20cookiecutter.__project_name_kebab%20%7D%7D/pyproject.toml

### Working on this

This needs a discussion with the core team. Once agreed, it's likely that this needs to be picked up by a core team member, considering how crucial it is to the package build and installation process.

**Comment by laymonage** [comment]
Related discussion on the pip repo: https://github.com/pypa/pip/issues/11457

**Comment by thibaudcolas** [comment]
If possible it’d be nice to preserve enough of the files to make use of the [GitHub dependency graph](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/dependency-graph-supported-package-ecosystems) features, as support doesn’t seem too good for `pyproject.toml` currently. 

See for example the [django-pattern-library setup.py](https://github.com/torchbox/django-pattern-library/blob/main/setup.py) which is there only to get picked up by this tooling.

**Comment by laymonage** [comment]
@thibaudcolas good point, from my experience that feature is very sensitive to those files and isn't very deterministic.

FWIW Django itself doesn't have any of the files mentioned in the docs but it seems to still work. Probably biased because of its popularity though, so maybe there's a manual entry on GitHub's side.

Here's a related page on the GitHub docs: https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/exploring-the-dependencies-of-a-repository#enabling-and-disabling-the-dependency-graph

With Wagtail, since we also have https://www.npmjs.com/package/wagtail, it shows up on the settings too:

<img width="1236" alt="Image" src="https://github.com/user-attachments/assets/3175e7f5-c2d9-44cb-a53e-810e7cf39536" />
